docs/run-tests-on-local.md

```diff
@@ -0,0 +1,116 @@
+# How To Run Functional Test And Integration Test On Local
+Run tests using python command or edgex-taf-common image depend on install the required libraries for robotframework on testing machine or not. If user don't like to install libraries for testing, the edgex-taf-common image is useful.
+### Prerequisites
+Clone the edgex-taf project from EdgeX Foundry as a template:
+``` bash
+cd ${HOME}
+git clone https://github.com/edgexfoundry/edgex-taf.git
+```
+
+### Using The edgeX-taf-common Container To Run Testing
+
+####  Variables configuration
+Export the following variables that depend on running environment
+``` 
+# Required variables
+export WORK_DIR=${HOME}/edgex-taf
+```
+
+#### Run test by shell script with arguments
+```
+cd ${WORK_DIR}/TAF/utils/script/docker
+sh run-tests.sh ${ARCH} ${SECURITY_SERVICE_NEEDED} ${TEST_STRATEGY}
+# ex. sh run-tests.sh x86_64 false 1
+
+# Variable values
+${ARCH}: x86_64 | arm64
+${SECURITY_SERVICE_NEEDED}: false | true
+${TEST_STRATEGY}: 1 (functional) | 2 (integration)
+```
+
+#### Test report
+Open the report file by browser: ${WORK_DIR}/TAF/testArtifacts/reports/cp-edgex/v2-api-test.html
+
+
+
+### Using Python Command To Run Testing
+#### Setup required library
+1. Install pre-request packages:
+    Download pip3 and run this command:
+    ``` bash
+    sudo apt-get install python3-pip
+    ```
+2. Install TAF common:
+    ``` bash
+    cd ${HOME}/edgex-taf
+    git clone https://github.com/edgexfoundry/edgex-taf-common.git
+    
+    # Install dependency lib
+    pip3 install -r ./edgex-taf-common/requirements.txt
+
+    # Install edgex-taf-common as lib
+    pip3 install ./edgex-taf-common
+    ```
+3. Prepare test environment:
+    ``` bash
+    # Fetch the latest docker-compose file
+    cd ${HOME}/edgex-taf/TAF/utils/scripts/docker
+    sh get-compose-file.sh ${USE_DB} ${ARCH} ${USE_SECURITY}
+    # ex. sh get-compose-file.sh -redis x86_64 -
+    
+    # Variables for get-compose-file.sh
+    ${USE_DB}: -redis | -mongo (mongo is not supported from hanoi release)
+    ${ARCH}: x86_64 | arm64
+    ${USE_SECURITY}: - (false) | -security- (true)
+
+    # export the environment variables that depend on your environment.
+    export ARCH=x86_64
+    export SECURITY_SERVICE_NEEDED=false
+    export COMPOSE_IMAGE=nexus3.edgexfoundry.org:10003/edgex-devops/edgex-compose:latest
+    ```
+#### Run Tests
+Run tests under ${HOME}/edgex-taf
+1. Deploy edgex:
+    ``` bash
+    # This step may take for a while if the edgex images don't exist on the machine
+    python3 -m TUC --exclude Skipped --include deploy-base-service -u deploy.robot -p default
```

Comment:
When I ran this, I got permission denied to remove the old files in edgex-taf/TAF/testArtifacts/reports/edgex.  I had to chmod the directory or run as sudo.  Each time I ran it, I had to change the permissions of these files again.

Reply:
Once I fixed the file permission problems, when I run the python script, I get the following:

```
jim@jim-ubuntu:~/forks/edgex-taf$ python3 -m TUC --exclude Skipped --include deploy-base-service -u deploy.robot -p default
INFO:root:Run testing for profile 'default'
INFO:root:Running use case None
==============================================================================
default :: Deploy EdgeX                                                       
==============================================================================
INFO:EdgeX deployment:**************************************************
INFO:EdgeX deployment:**************************************************
INFO:EdgeX deployment:************  DEMO TEST RUN START  ************
INFO:EdgeX deployment:Suite Setup: Logging started, level: INFO
INFO:EdgeX deployment:Suite Setup: python version: 3.6.9 (default, Oct  8 2020, 12:12:24) 
[GCC 8.4.0]
INFO:EdgeX deployment:Suite Setup: testcases version: 1.0
Deploy EdgeX                                                          INFO:EdgeX deployment:Deploy EdgeX
INFO:EdgeX deployment:unknown shorthand flag: 'f' in -f

INFO:EdgeX deployment:See 'docker run --help'.

INFO:EdgeX deployment:exit 0
INFO:EdgeX deployment:Success to execute cmd: sh /home/jim/forks/edgex-taf/TAF/utils/scripts/docker/deploy-edgex.sh
INFO:EdgeX deployment:Check service data is startup...
INFO:EdgeX deployment:Ping service with port 48080 and request url localhost /api/v1/ping ... 
Deploy EdgeX                                                          | FAIL |
ConnectionRefusedError: [Errno 111] Connection refused
------------------------------------------------------------------------------
INFO:EdgeX deployment:Suite teardown
INFO:EdgeX deployment:************    DEMO TEST RUN END  ************
INFO:EdgeX deployment:**************************************************
INFO:EdgeX deployment:**************************************************
default :: Deploy EdgeX                                               | FAIL |
1 critical test, 0 passed, 1 failed
1 test total, 0 passed, 1 failed
==============================================================================
Output:  /home/jim/forks/edgex-taf/TAF/testArtifacts/reports/edgex/report.xml
Log:     /home/jim/forks/edgex-taf/TAF/testArtifacts/reports/edgex/log.html
Report:  /home/jim/forks/edgex-taf/TAF/testArtifacts/reports/edgex/report.html
```

Reply:
When I just try to bring up EdgeX with the docker compose file referenced, I saw two things:
1 - an existing running mosquitto instance had to be stopped or else the port was colliding
2 - I had to clean out all existing volumes
Even then, it would not start EdgeX device virtual or device modbus.  

```
jim@jim-ubuntu:~/forks/edgex-taf/TAF/utils/scripts/docker$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
jim@jim-ubuntu:~/forks/edgex-taf/TAF/utils/scripts/docker$ docker-compose up -d
WARNING: The DOCKER_HOST_IP variable is not set. Defaulting to a blank string.
WARNING: The MQTT_BROKER_IP variable is not set. Defaulting to a blank string.
WARNING: The CONF_DIR variable is not set. Defaulting to a blank string.
WARNING: The PROFILE variable is not set. Defaulting to a blank string.
Building with native build. Learn about native build in Compose here: https://docs.docker.com/go/compose-native-build/
Creating network "docker_edgex-network" with driver "bridge"
Creating volume "docker_consul-config" with default driver
Creating volume "docker_consul-data" with default driver
Creating volume "docker_db-data" with default driver
Creating volume "docker_kuiper-data" with default driver
Creating edgex-modbus-simulator      ... done
Creating edgex-mqtt-broker           ... done
Creating edgex-core-consul      ... done
Creating edgex-redis            ... done
Creating edgex-support-scheduler     ... done
Creating edgex-support-notifications ... done
Creating edgex-core-metadata         ... done
Creating edgex-core-data             ... done
Creating edgex-core-command          ... done
Creating app-service-functional-tests ... 
Creating edgex-sys-mgmt-agent         ... 
Creating app-service-mqtt-export      ... 
Creating edgex-device-virtual                 ... error
Creating edgex-device-rest            ... 
Creating edgex-device-modbus          ... 
Creating app-service-functional-tests         ... done
Creating edgex-sys-mgmt-agent                 ... done
Creating app-service-mqtt-export              ... done

Creating edgex-device-rest                    ... done
z': invalid mount config for type "bind": invalid mount path: '.' mount path must be absolute
Creating app-service-http-export              ... done
Creating edgex-app-service-configurable-rules ... done
Creating scalability-test-mqtt-export         ... done
Creating edgex-kuiper                         ... done

ERROR: for device-virtual  Cannot create container for service device-virtual: invalid volume specification: '/home/jim/forks/edgex-taf/TAF/config:.:z': invalid mount config for type "bind": invalid mount path: '.' mount path must be absolute

ERROR: for edgex-device-modbus  Cannot create container for service edgex-device-modbus: invalid volume specification: '/home/jim/forks/edgex-taf/TAF/config:.:z': invalid mount config for type "bind": invalid mount path: '.' mount path must be absolute
ERROR: Encountered errors while bringing up the project.
```

Reply:
> Once I fixed the file permission problems, when I run the python script, I get the following:
> 
> ```
> jim@jim-ubuntu:~/forks/edgex-taf$ python3 -m TUC --exclude Skipped --include deploy-base-service -u deploy.robot -p default
> INFO:root:Run testing for profile 'default'
> INFO:root:Running use case None
> ==============================================================================
> default :: Deploy EdgeX                                                       
> ==============================================================================
> INFO:EdgeX deployment:**************************************************
> INFO:EdgeX deployment:**************************************************
> INFO:EdgeX deployment:************  DEMO TEST RUN START  ************
> INFO:EdgeX deployment:Suite Setup: Logging started, level: INFO
> INFO:EdgeX deployment:Suite Setup: python version: 3.6.9 (default, Oct  8 2020, 12:12:24) 
> [GCC 8.4.0]
> INFO:EdgeX deployment:Suite Setup: testcases version: 1.0
> Deploy EdgeX                                                          INFO:EdgeX deployment:Deploy EdgeX
> INFO:EdgeX deployment:unknown shorthand flag: 'f' in -f
> 
> INFO:EdgeX deployment:See 'docker run --help'.
> 
> INFO:EdgeX deployment:exit 0
> INFO:EdgeX deployment:Success to execute cmd: sh /home/jim/forks/edgex-taf/TAF/utils/scripts/docker/deploy-edgex.sh
> INFO:EdgeX deployment:Check service data is startup...
> INFO:EdgeX deployment:Ping service with port 48080 and request url localhost /api/v1/ping ... 
> Deploy EdgeX                                                          | FAIL |
> ConnectionRefusedError: [Errno 111] Connection refused
> ------------------------------------------------------------------------------
> INFO:EdgeX deployment:Suite teardown
> INFO:EdgeX deployment:************    DEMO TEST RUN END  ************
> INFO:EdgeX deployment:**************************************************
> INFO:EdgeX deployment:**************************************************
> default :: Deploy EdgeX                                               | FAIL |
> 1 critical test, 0 passed, 1 failed
> 1 test total, 0 passed, 1 failed
> ==============================================================================
> Output:  /home/jim/forks/edgex-taf/TAF/testArtifacts/reports/edgex/report.xml
> Log:     /home/jim/forks/edgex-taf/TAF/testArtifacts/reports/edgex/log.html
> Report:  /home/jim/forks/edgex-taf/TAF/testArtifacts/reports/edgex/report.html
> ```

Did you export the following variable before executing this command?
```
# export the environment variables that depend on your environment.
    export ARCH=x86_64
    export SECURITY_SERVICE_NEEDED=false
    export COMPOSE_IMAGE=nexus3.edgexfoundry.org:10003/edgex-devops/edgex-compose:latest
```

Reply:
> When I ran this, I got permission denied to remove the old files in edgex-taf/TAF/testArtifacts/reports/edgex. I had to chmod the directory or run as sudo. Each time I ran it, I had to change the permissions of these files again.

The reason is we use root to run tests in docker container, and the script will create a report folder by root. We suggest run tests either docker container or python command. I will add a note to the document.

Reply:
> When I just try to bring up EdgeX with the docker compose file referenced, I saw two things:
> 1 - an existing running mosquitto instance had to be stopped or else the port was colliding
> 2 - I had to clean out all existing volumes
> Even then, it would not start EdgeX device virtual or device modbus.
> 

The docker-compose file contains some variables that need to get from TAF scripts, so we don't suggest starting services directly by docker-compose up.